### DISCUSSION TITLE
Merging data from multiple sources during import

### DISCUSSION THREAD

**Discussion by snejus** [beetbox/beets#5422] [Ideas]
Currently, when we import an album we need to select one of the available options to import, where in most cases we have different combinations of *data_source* and *media* fields.

In those cases where an album is found in multiple databases/platforms, we can only select one of them, for example:

![image](https://github.com/user-attachments/assets/5f605492-f8d9-443c-9ae3-3ba1e8a2ddf6)

Here, **Digital Media** media of this album is available on MusicBrainz, Discogs and Bandcamp. While most fields are set by each source, there are some that are only set by one of the sources, see below

#### MusicBrainz
![image](https://github.com/user-attachments/assets/dbd9c0f3-1dfe-4e8e-8c53-4e9dfc6e3ffb)

#### Discogs
![image](https://github.com/user-attachments/assets/de3dd136-d4bf-44ca-b561-745144dcb12b)

#### Bandcamp
![image](https://github.com/user-attachments/assets/18790f8f-e30a-45a2-a670-854e5ec3beb6)
![image](https://github.com/user-attachments/assets/c772fb8a-0c5f-4fdc-835f-4501ee4e14f7)

For example, only MusicBrainz sets all `artist_sort` fields, only Discogs has field `catalognum` available while Bandcamp has `comments`. Ideally, I would like to have these merged.

What do we think about this?

A couple of initial thoughts regarding a potential MVP implementation:

1. A configuration option `source_priority`, e.g.
   ```yaml
   source_priority: ["MusicBrainz", "bandcamp"]
   ```

2. Once we've got release candidates, we can group them by **media** and, say, **album**
3. The first available release from the `source_priority` configuration becomes the base release
4. If there are more releases in the group, we grab from them every field that is not set in the base release

**Comment by arsaboo** [comment]
This is exactly what I have been after (I even started working on something like [meta-import](https://github.com/arsaboo/beets-metaimport), but did not get too far). Now that we have multiple tag sources, we SHOULD be able to import all the relevant fields from multiple sources and use them all. 

I like the idea of `source_priority`, which can be used to break ties. For other fields unique to each source (e.g., `spotify_popularity`), we just import them all.

**Comment by dogo77** [comment]
> for example:
> 
> ![image](https://private-user-images.githubusercontent.com/16212750/367603896-5f605492-f8d9-443c-9ae3-3ba1e8a2ddf6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3NDE5NTcsIm5iZiI6MTcyNjc0MTY1NywicGF0aCI6Ii8xNjIxMjc1MC8zNjc2MDM4OTYtNWY2MDU0OTItZjhkOS00NDNjLTlhZTMtM2JhMWU4YTJkZGY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE5VDEwMjczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2ZWZjYzA3ZTgxZDA5ZjgwMmNiODkyNWJkOThhNDM1ZTlkY2YzN2Y2MDI3MDliNjExMmUyYTdhNWRmNDc3NWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vYTjoxl8g7W6iPvqMeSWML2FAwG-rRam95ZHbCLL_wY)
> 
sorry, but may i ask: 
HOW do you get this UI???
Running beets-git in arch definitely looks different.


And on Topic: +1 for the idea of merging different information providers while importing ;)

**Reply by snejus** [reply]
> sorry, but may i ask: HOW do you get this UI??? Running beets-git in arch definitely looks different.

You can find this on my fork, [`dev` branch](https://github.com/beetbox/beets/compare/master...snejus:beets:dev)

There's no documentation however, and it's very likely to fail on your end because it depends on several configuration updates. Feel free to message me on Telegram (see my GitHub profile for the link) - I'll be happy to hear what breaks and needs fixing.
